src/os/filestore/FileStore.cc

```diff
@@ -1794,11 +1806,249 @@ int FileStore::umount()
+    error = thr->Start();
+    if (error)
+      break;
+  }
+
+  if (error) {
+    derr << "unable to create read completion threads, error = "
+	 << error << dendl;
+    for (auto rct : read_completion_threads)
+      delete rct;
+    read_completion_threads.clear();
+  }
+ 
+  return error;
+}
+
+void FileStore::destroy_read_completion_threads()
+{
+  if (read_completion_threads.size() == 0)
+    return;
+  dout(0) << "FileStore: destroying " << read_completion_threads.size()
+	  << " read completion threads." << dendl;
+
+  for (auto rct : read_completion_threads)
+    delete rct;
+
+  dout(0) << "FileStore:  destroy_read_completion_threads done" << dendl;
+  read_completion_threads.clear();
+}
+
+int FileStore::async_read_dispatch(
+  Sequencer *posr,
+  Context *ctx,
+  const coll_t& cid,
+  const ghobject_t& oid,
+  uint64_t off,
+  size_t len,
+  bufferlist *bl,
+  uint32_t op_flags,
+  Context *blessed_context)
+{
+  OpSequencer *osr = may_create_op_sequencer(posr);
+  // PGID-to-completion thread mapping:
+  int thread_index = osr->id % read_completion_threads.size();
+
+  FDRef fd;
+  off64_t offset = (off64_t) off;
+  int r = lfn_open(cid, oid, false, &fd, 0, true);
+  if (r < 0) {
+    dout(10) << "FileStore::async_read_dispatch(" << cid << "/" << oid << ") open error: "
+	     << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  if (len == 0) {
+    struct stat st;
+    memset(&st, 0, sizeof(struct stat));
+    int r = ::fstat(**fd, &st);
+    assert(r == 0);
+    len = st.st_size;
+  }
+
+#ifdef HAVE_POSIX_FADVISE
+  if (op_flags & CEPH_OSD_OP_FLAG_FADVISE_RANDOM)
+    posix_fadvise(**fd, offset, len, POSIX_FADV_RANDOM);
+  if (op_flags & CEPH_OSD_OP_FLAG_FADVISE_SEQUENTIAL)
+    posix_fadvise(**fd, offset, len, POSIX_FADV_SEQUENTIAL);
+#endif
+
+  // Extend to 512 bytes since direct IO is being used
+  unsigned alignment = len % 512;
+  if (alignment)
+    alignment = 512 - alignment;
+
+  read_aio_bl(fd, bl, offset, len, alignment, ctx, blessed_context, op_flags, oid, thread_index);
+  return r;
+}
+
+void FileStore::read_aio_bl(FDRef fdref, bufferlist *bl, off64_t& pos,
+                            size_t len, unsigned alignment, Context *ctx, Context *pctx,
+                            uint32_t op_flags, const ghobject_t& oid, int thread_index)
+{
+  int fd = **fdref;
+  read_aio *rio = new read_aio(fdref, bl, pos, len, alignment, op_flags, oid, ctx, pctx);
 
+  len += alignment;
+  io_prep_pread(&rio->iocb, fd, rio->bptr->c_str(), len, pos);
+  dout(20) << "read_aio_bl .. " << len << dendl;
+
+  ReadCompletionThread& rct = *read_completion_threads[thread_index];
+  // Error if thread no longer up.
+  int status = rct.inc_thrd_pending();
+  if (status < 0) {
+    rio->len = status;
+    async_read_finish(rio);
+    return;
+  }
+
+  io_context_t *iocontext = &rct.read_aio_ctx;
+
+  iocb *piocb = &rio->iocb;
+  int attempts = 10;
+  do {
+    int r = io_submit(*iocontext, (long) 1, &piocb);
```

Comment:
io_submit can block for arbitrary time for ordinary files. The reason why it blocks is because most filesystems perform blocking disk operations internally, and as long as jobs are submitted to read and write files in a filesystem, they will spend a lot of time blocking during submit.
This is not the case when submitting jobs to block devices (opened with O_DIRECT).

Are you sure it's OK to block the calling thead here?